src/sync/spsc_bufferqueue.rs

```diff
@@ -0,0 +1,400 @@
+//! SPSC ringbuffer
+
+use std::sync::atomic::Ordering::{Acquire, Release, Relaxed, SeqCst};
+use std::sync::atomic::{AtomicUsize, AtomicBool, fence};
+use std::sync::Arc;
+use std::ptr;
+use std::mem;
+use std::cmp;
+use std::marker::PhantomData;
+use mem::CachePadded;
+
+#[inline(always)]
+unsafe fn deallocate<T>(ptr: *mut T, old_size: usize) {
+    Vec::from_raw_parts(ptr, 0, old_size);
+}
+
+#[inline(always)]
+unsafe fn allocate<T>(size: usize) -> *mut T {
+    let mut buf = Vec::with_capacity(size);
+    let ptr = buf.as_mut_ptr();
+    mem::forget(buf);
+    ptr
+}
+
+/// A single-producer, single consumer bounded wait-free ringbuffer queue
+///
+/// All operations on the buffer queue are wait-free,
+/// provided move/copy operations are waitfree.
+/// This queue does not allocate after constructions
+#[repr(C)]
+pub struct SpscBufferQueue<T: Send> {
+    // This is a pointer instead of a vector
+    // so that vector doesn't call constructors
+    data_block: *mut T,
+    size: usize,
+    _marker: PhantomData<T>,
+
+    // These dummies result in a tremendous performance improvement, ~300%+
+    _dummy_1: CachePadded<u64>,
```

Comment:
I thought about that, but am not sure if it's worth the more complicated datastructure definition/member functions. It would get rid of the drop warning, but that could also be accomplished with a single other datastructure wrapped by the dropping one.

I can see this getting harder to scale to a larger or more complicated object though.